client/src/main/java/io/pravega/client/stream/impl/ControllerImpl.java

```diff
@@ -254,38 +258,42 @@ public ControllerImpl(ManagedChannelBuilder<?> channelBuilder) {
+                    final long start = System.currentTimeMillis();
+                    return FutureHelpers.loop(() -> !done.get(), () ->
+                            FutureHelpers.delayedFuture(() -> checkScaleStatus(stream, startScaleResponse.getEpoch())
+                                    .thenAccept(isDone -> {
+                                        done.set(isDone);
+                                        if (!isDone) {
+                                            if (System.currentTimeMillis() > (start + timeoutMillis)) {
+                                                throw new ControllerFailureException("scale request timed out");
+                                            }
+                                        }
+                                    }), 1000, executor), executor)
+                            .thenApply(x -> started);
+                });
+    }
+
+    @Override
+    public CompletableFuture<Boolean> startScale(final Stream stream, final List<Integer> sealedSegments,
                                                   final Map<Double, Double> newKeyRanges) {
         long traceId = LoggerHelpers.traceEnter(log, "scaleStream", stream);
-        Preconditions.checkNotNull(stream, "stream");
-        Preconditions.checkNotNull(sealedSegments, "sealedSegments");
-        Preconditions.checkNotNull(newKeyRanges, "newKeyRanges");
-
-        RPCAsyncCallback<ScaleResponse> callback = new RPCAsyncCallback<>();
-        client.scale(ScaleRequest.newBuilder()
-                             .setStreamInfo(ModelHelper.createStreamInfo(stream.getScope(), stream.getStreamName()))
-                             .addAllSealedSegments(sealedSegments)
-                             .addAllNewKeyRanges(newKeyRanges.entrySet().stream()
-                                                         .map(x -> ScaleRequest.KeyRangeEntry.newBuilder()
-                                                                 .setStart(x.getKey()).setEnd(x.getValue()).build())
-                                                         .collect(Collectors.toList()))
-                             .setScaleTimestamp(System.currentTimeMillis())
-                             .build(),
-                     callback);
-        return callback.getFuture().thenApply(x -> {
+        return startScaleInternal(stream, sealedSegments, newKeyRanges).thenApply(x -> {
             switch (x.getStatus()) {
             case FAILURE:
                 log.warn("Failed to scale stream: {}", stream.getStreamName());
                 throw new ControllerFailureException("Failed to scale stream: " + stream);
             case PRECONDITION_FAILED:
                 log.warn("Precondition failed for scale stream: {}", stream.getStreamName());
                 return false;
-            case SUCCESS:
+            case STARTED:
                 log.info("Successfully scaled stream: {}", stream.getStreamName());
```

Comment:
We need to change this log info message because at this point the scale isn't complete yet.